internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,385 @@
+package jobobject
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/processorinfo"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+// This file provides higher level constructs for the win32 job object API.
+// Most of the core creation and management functions are already present in "golang.org/x/sys/windows"
+// (CreateJobObject, AssignProcessToJobObject, etc.) as well as most of the limit information
+// structs and associated limit flags. Whatever is not present from the job object API
+// in golang.org/x/sys/windows is located in /internal/winapi.
+//
+// https://docs.microsoft.com/en-us/windows/win32/procthread/job-objects
+
+// JobObject is a high level wrapper around a Windows job object. Holds a handle to
+// the job, a channel to receive iocp notifications about the lifecycle
+// of the job and a mutex for synchronized handle access.
+type JobObject struct {
+	jobHandle  windows.Handle
+	notifChan  chan interface{}
+	handleLock sync.RWMutex
+}
+
+type jobLimits struct {
+	cpuCount        uint32
+	cpuRate         uint32
+	cpuWeight       uint32
+	memoryLimitInMB uintptr
+	maxIops         int64
+	maxBandwidth    int64
+}
+
+// Processor resource controls
+const (
+	processorLimitMin  = 1
+	processorLimitMax  = 10000
+	processorWeightMin = 0
+	processorWeightMax = 10000
+)
+
+var errAlreadyClosed = errors.New("the handle has already been closed")
+
+// SetResourceLimits sets resource limits on the job object (cpu, memory, storage).
+func (job *JobObject) SetResourceLimits(ctx context.Context, limits *jobLimits) error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	// Go through and check what limits were specified and apply them to the job.
+	if limits.memoryLimitInMB != 0 {
+		var (
+			basicLimitFlags uint32
+			eliInfo         windows.JOBOBJECT_EXTENDED_LIMIT_INFORMATION
+		)
+		basicLimitFlags |= windows.JOB_OBJECT_LIMIT_JOB_MEMORY
+		if (limits.memoryLimitInMB << 20) > math.MaxUint64 {
```

Comment:
I think we should just take a memory limit in bytes, and avoid this shifting.

Reply:
The problem was I was re-using `oci.ParseAnnotationsMemory` in the container work which returns either the annotation memory amount specified in MB or the oci limit converted to MB. So I'd either just have to take that and reconvert to bytes anyways

Reply:
I'd encourage you to think about this package as supporting job objects, without any focus on serving a particular container implementation. I'd like to expose the direct platform capabilities as much as reasonable, so taking a bytes value here seems appropriate to me.

In this case I think doing the conversion in the caller is the right place for it.

Reply:
As an example of why this could be important, there's a good chance we will want to use this package to assign job object limits on the shim itself at some point. So having a good job object interface that isn't focused on containers will be really helpful.